keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1194 @@
+   - Waiting for a cleanup and release of important resources held by the pod.
+   - Waiting for important computations to complete.
+   - Non-graceful deletion of the pod (`gracePeriodSeconds=0`).
+   - Deletion of a pod that is covered by a blocking PodDisruptionBudget. The controller of the
+     application should have additional logic to distinguish whether a disruption of a particular
+     pod will disrupt the application as a whole.
+
+In the end, the evacuation should always end with a pod being deleted by either the evacuator or
+via an eviction by the evacuation controller.
+
+We should discourage the creation of preventive evacuations, so that they do not end up as
+another PDB. So we should design the API appropriately and also not allow behaviors that do not
+conform to the evacuation contract.
+
+### Evacuation controller
+
+In order to fully enforce the evacuation contract and prevent code duplication among evacuation
+instigators, we will introduce a new controller called the evacuation controller.
+
+Its responsibility is to observe evacuation requests from instigators and periodically check that
+evacuators are making progress in evacuating evacuees. It is important to see a consistent effort
+by the evacuators to reconcile the progress of the evacuation. This is important to prevent stuck
+evacuations that could bring node maintenance to a halt. If the evacuation controller detects
+that the evacuation progress updates have stopped (or never started), it will resort to pod
+eviction by calling the eviction API (taking PodDisruptionBudgets into consideration).
+
+It is also responsible for garbage collection/deletion of existing evacuations whose pods have
+already been deleted.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin I want high-level components like node maintenance (planned replacement of
+kubectl drain), scheduler, descheduler to use the Evacuation API to gracefully remove pods from a
+set of nodes. I also want to see the progress of ongoing evacuations and be able to debug them if
+something goes wrong. This means to:
+- Easily identify pods that have accepted evacuation and are making progress. If possible to be
+  able to see evacuation's ETA.
+- Identify pods that should be evicted instead of evacuated and to distinguish pods that are
+  failing eviction.
+- See additional debug information from the evacuator and to be able to identify the evacuator.
+
+#### Story 2
+
+As an application owner, I want to run single replica applications without disruptions and have the
+ability to easily migrate the workload pods from one node to another. This also applies to
+applications with larger number of replicas that prefer to surge (upscale) pods first instead of
+downscaling.
+
+### Notes/Constraints/Caveats (Optional)
+
+DaemonSet pods and mirror pods can be evacuated, but cannot be evicted because they can run
+critical services.
```

Comment:
What does it mean to evacuate a static pod, or its mirror Pod?

Reply:
I have added some details to highlight the future use cases in these areas.